frontend: update 'My portfolio' with lightning account

strmci opened the pull request:
This commit adds functionality to display Lightning account
in 'My portfolio' in cases where it is active but not associated
with a connected or remembered wallet, or when all mainnet
accounts in the connected wallet are disabled.

The total coins table is updated to include the Lightning account
Hide chart, percentage change, and filters in my portfolio if
only the lightning account is displayed.

strmci commented on frontends/web/src/routes/account/summary/accountssummary.tsx:
```diff
@@ -60,9 +61,28 @@ export function AccountsSummary({ accounts, devices }: TProps) {
+    } else {
+      return false;
+    }
+  };
```
nice, I just had to change the `rootFingerprint` condition to 
` return keystore.rootFingerprint !== lightningConfig.accounts[0].rootFingerprint;`
and it works fine, thanks

thisconnect commented on frontends/web/src/routes/account/summary/accountssummary.tsx:
```diff
@@ -213,13 +228,16 @@ export function AccountsSummary({ accounts, devices }: TProps) {
             )}
+            {hasLightningFromOtherKeystore && (
+              <LightningBalance/>
+            )}
```
is my understanding correct, that if the Lightning account is from a connected keystore it will be included together with normal accounts?

strmci replied:
yes, it will be included together with normal accounts

strmci replied:
it will be included with normal accounts also when it is from a remembered keystore

thisconnect commented on frontends/web/src/routes/account/summary/coinbalance.tsx:
```diff
@@ -20,6 +20,8 @@ import { SubTotalCoinRow } from './subtotalrow';
 import { Skeleton } from '../../../components/skeleton/skeleton';
 import style from './accountssummary.module.css';
+import { useLightning } from '../../../hooks/lightning';
+import { CoinCode } from '../../../api/account';
```
please always import styles at the end.

strmci replied:
Updated.

thisconnect commented on frontends/web/src/routes/account/summary/coinbalance.tsx:
```diff
@@ -50,6 +56,22 @@ export function CoinBalance ({
   const accountsPerCoin = getAccountsPerCoin();
   const coins = Object.keys(accountsPerCoin) as accountApi.CoinCode[];
 
+  let coinsWithLightning: TCoinMap = {} as TCoinMap;
```
If the intention is to re-assign the variable later use `let` else better use `const`.

strmci replied:
I wasn't sure here, updated.

thisconnect commented on frontends/web/src/routes/account/summary/coinbalance.tsx:
```diff
@@ -70,21 +92,18 @@ export function CoinBalance ({
+                  />
+                );
+              }
+              )) : null}
```
this looks incorrect:


<img width="1238" alt="Screenshot 2024-05-10 at 09 19 18" src="https://github.com/digitalbitbox/bitbox-wallet-app/assets/546900/dc98bff3-b408-4c6c-9050-10c0948a203f">

strmci replied:
it is correct, you have Bitcoin (in lightning) and also some bitcoin testnet coins.

but it could be sorted better as:
Bitcoin
Bitcoin Testnet
...

thisconnect commented on frontends/web/src/routes/account/summary/coinbalance.tsx:
```diff
@@ -70,21 +92,18 @@ export function CoinBalance ({
-                return null;
-              })) : null}
+            { (Object.keys(coinsWithLightning).length > 0) ? (
+              Object.entries(coinsWithLightning).map(([coinCode, coinName]) => {
```
When reading this I could understand what the intention was (iterative over coins with at least 1 account or more), but now I find it hard to understand what the intention is.

strmci replied:
the intention is to iterate over `coinsWithLightning` but only if there are any keys in it so it won't iterate if it is empty

thisconnect commented on frontends/web/src/routes/account/summary/coinbalance.tsx:
```diff
@@ -50,6 +56,22 @@ export function CoinBalance ({
+    if (!coinsWithLightning[bitcoinCode]) {
+      coinsWithLightning[bitcoinCode] = bitcoinName;
+    }
+  }
```
coinsWithLightning is confusing me, but maybe it's only the name?

I find it hard to read. It creates a map here, and then later turns the map into Object.keys and Object.entries. We already receave another map in this component coinsBalances and have a helped function to create another map accountsPerCoin.. feel like this all could be simplified, but I am not sure.

Couldn't you just make a `const totals = [];` array instead of the map and push coins/total in there somehow? 

Also we already have `coinsBalances` map 🤔

strmci replied:
Yeah, the name may be confusing.

eg. when you have only a Litecoin account enabled the `accountsPerCoin` will only contain this account, so I created `coinsWithLightning`to add Bitcoin there (when Lightning is active and if `accountsPerCoin` doesn't include it yet)

`coinsBalances` can't be used as they contain just balances, not active accounts. If the amounts are 0 on the accounts, we want to show them. It also doesn't include `coinName`.

strmci replied:
yeah, you are right, it can be simplified to use only `coinsBalances`. I have to include the coin name there.

strmci replied:
We miss the coin name in `coinsBalances`...

Beerosagos replied:
I agree, this is kind of confusing to me, too. Also, I think that if some info is missing, it should come from the backend in a simple and straight-forward way, as a general approach.

thisconnect replied:
this OR simplifies the logic and should come first: https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2714

thisconnect replied:
ok there is some nice stuff in my PR, but the coinName and order of coins/tokens shouldn't be job of the frontend. So I will close the PR, but maybe take some TS improvements and make a separate PR.

Beerosagos commented on frontends/web/src/routes/account/summary/accountssummary.tsx:
```diff
@@ -60,9 +61,23 @@ export function AccountsSummary({ accounts, devices }: TProps) {
+  const hasLightningFromOtherKeystore = (
+    lightningConfig.accounts.length !== 0
+    && (
+      accountsByKeystore.length === 0
```
I think that you can avoid checking the length, it should be handled by the some condition anyway, right?

strmci replied:
Yes, this check is redundant here, removed.

Beerosagos commented on frontends/web/src/routes/account/summary/accountssummary.tsx:
```diff
@@ -60,9 +61,23 @@ export function AccountsSummary({ accounts, devices }: TProps) {
+    )
+  );
+
+  const showTotalCoins = accountsByKeystore.length > 1 || (hasLightningFromOtherKeystore && accountsByKeystore.length === 1);
```
Nit: what about doing smth like

`let keystores = accountsByKeystore.length;
if (hasLightningFromOtherKeystore) {
    keystores += 1;
}`
  
  and check for keystores > 1 below?
  
  I feel like it would be more readable

strmci replied:
I updated this to use `keystores`

Beerosagos commented:
Here I found the keystore (c6d..) repeated twice, during my test

![image](https://github.com/BitBoxSwiss/bitbox-wallet-app/assets/11994448/a5742f6f-ebff-4408-9f3f-b7e9edd9d773)

strmci commented on frontends/web/src/routes/account/summary/summarybalance.tsx:
```diff
@@ -166,3 +166,51 @@ export function SummaryBalance({
+    <div>
+      <div className={style.accountName}>
+        <p>
+          {lightningConfig.accounts[0].keystoreName} ({lightningConfig.accounts[0].rootFingerprint})
```
please try to disable and enable the lightning wallet on this branch again

strmci commented:
@Beerosagos 
I cherry-picked the total coins refactor from master and updated this PR.

> Here I found the keystore (c6d..) repeated twice, during my test
Also, this should be resolved now.

PTAL

thisconnect commented:
Maybe merge master->staging-ln before this?

Beerosagos commented:
> There was some refactoring on master, maybe consider merge master->staging-ln before this?

Yes, I agree :+1:

Beerosagos commented:
merge master > staging-ln PR: https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2782

strmci commented:
@Beerosagos rebased.
PTAL

Beerosagos commented on backend/config/lightning.go:
```diff
@@ -26,6 +26,8 @@ type LightningAccountConfig struct {
 	Mnemonic string `json:"mnemonic"`
 	// RootFingerprint is fingerprint of the keystore that generated the entropy.
 	RootFingerprint jsonp.HexBytes `json:"rootFingerprint"`
+	// KeyStoreName is name of the keystore.
```
I realized that storing the keystore name could lead to issues when the device name is changed by the user.
I didn't look deeply into this, but I think that the proper way of doing it could be to get the persisted keystore info and retrieve the name there (see `backend.Config().AccountsConfig()`), but we may need to actively persist the current keystore when enabling the LN account).

cc @benma wdyt?

Beerosagos reviewed (changes requested):
Left a couple comments.

Also, I noticed that if the only active account is the lightning one, then the chart in the account summary is empty. Not sure what we want to do with this, the easier is probably to hide it, at least for now. wdyt @thisconnect ?

thisconnect commented:
> Also, I noticed that if the only active account is the lightning one, then the chart in the account summary is empty. Not sure what we want to do with this, the easier is probably to hide it, at least for now. wdyt @thisconnect ?

ack to hiding the chart for now if there is only lightning. But it would be nice in a future PR to have the chart also is there if there is only the lightining account.

strmci commented:
@Beerosagos 
- as discussed, moved the addition of the lightning balance after the accounts iteration and added Bitcoin to the beginning of `sortedCoins` if it is not already there
- hid the chart in my portfolio if only the lightning account is active
- updated to use `Config().AccountsConfig().LookupKeystore` to get the lightning account keystore name
  
Currently, we use the `isAmbiguousName` function to determine the table title for watch-only accounts. We could apply a similar approach for lightning accounts from other keystores. For now, if there’s an active lightning account from another keystore, the table title will be displayed as "keystoreName (rootFingerprint)".

PTAL

Beerosagos commented on backend/handlers/handlers.go:
```diff
@@ -1543,10 +1559,16 @@ func (handlers *Handlers) getLightningConfigHandler(_ *http.Request) interface{}
+		keystoreName := ""
+		lightningKeystore, err := handlers.backend.Config().AccountsConfig().LookupKeystore(account.RootFingerprint)
+		if err == nil {
+			keystoreName = lightningKeystore.Name
```
This is an interesting solution, but I think it has a defect: if the keystore device is renamed, the config in the frontend is not updated bc it is served by the provider. That means the keystore name won't be updated until the app is restarted. I think it would be better to add a backend endpoint `keystore-name` that given a rootfingerprint looks for a keystore in the config and returns the name. 

There is also a more latent bug, which depends on the watch-only implementation: when the device is renamed, the keystore name in the config file is not updated until the device is disconnected and reconnected again. But this deserves its own PR as it is out of scope here, imo. 

cc @benma

Beerosagos commented on frontends/web/src/routes/account/summary/summarybalance.tsx:
```diff
@@ -158,3 +158,55 @@ export const SummaryBalance = ({
 };
+
+
+export function LightningBalance() {
```
There is some code duplication here, I think that it would be nice to move the logic to subscribe and fetch the LN balance inside accountsummary.tsx and pass it to the LightningBalance and SummaryBalance components. wdyt?

Beerosagos commented:
Also, I noticed that when the chart is hidden, the `week/month/..` filter buttons are still visible (disabled). They should be hidden too, I think.

Beerosagos commented:
> Currently, we use the `isAmbiguousName` function to determine the table title for watch-only accounts. We could apply a similar approach for lightning account from other keystore. For now, if there’s an active lightning account from another keystore, the table title will be displayed as "keystoreName (rootFingerprint)".

Yeah, I agree :+1:

strmci commented:
@Beerosagos
- moved the logic to subscribe and fetch the LN balance inside accountsummary.tsx and pass it to the LightningBalance and SummaryBalance components
- created a new endpoint to get the keystore name from the root fingerprint
- refactored the `isAmbiguousName` function, so it is easier to use with lightning account (already in master: https://github.com/BitBoxSwiss/bitbox-wallet-app/commit/661f819cda87ea5e7087a5b9f746e6f60addbc48)
  and updated how it is used in the account summary
- hid the filter buttons together with the chart and percentage difference when only a lightning account is active

PTAL

Beerosagos commented on frontends/web/src/api/account.ts:
```diff
@@ -440,6 +440,15 @@ export const connectKeystore = (code: AccountCode): Promise<{ success: boolean;
   return apiPost(`account/${code}/connect-keystore`);
 };
 
+export type TResponseKeystoreName = null | {
```
why `null | `? I think you can remove it

Beerosagos commented on backend/handlers/handlers.go:
```diff
@@ -216,6 +216,7 @@ func NewHandlers(
 	getAPIRouterNoError(apiRouter)("/testing", handlers.getTesting).Methods("GET")
 	getAPIRouterNoError(apiRouter)("/account-add", handlers.postAddAccount).Methods("POST")
 	getAPIRouterNoError(apiRouter)("/keystores", handlers.getKeystores).Methods("GET")
+	getAPIRouterNoError(apiRouter)("/keystore-name", handlers.getKeystoreName).Methods("POST")
```
Why POST method? We usually use POST for calls that change data or that actively do something. In this case, since you are getting data instead of modifying it you should use a GET

Beerosagos commented on frontends/web/src/routes/account/summary/accountssummary.tsx:
```diff
@@ -65,9 +68,61 @@ export const AccountsSummary = ({
+    const lightningAccounts = lightningConfig.accounts;
+    if (lightningAccounts.length) {
+      fetchLightningBalance();
+      const subscriptions = [subscribeNodeState(fetchLightningBalance)];
```
When there is a node state change you should also fetch the coinsTotalBalance again, otherwise you miss the update there

Beerosagos commented on frontends/web/src/routes/account/summary/accountssummary.tsx:
```diff
@@ -65,9 +68,61 @@ export const AccountsSummary = ({
+  let keystores = accountsByKeystore.length;
+  let lightningAccountKeystoreName = '';
+  let keystoreDisambiguatorName;
+  if (hasLightningFromOtherKeystore) {
```
This whole `if` condition is not very readable, IMO. Can you try to simplify a bit? Maybe it could be useful to return an empty string instead of `null` in the `lightningKeystoreName` useLoad?

Beerosagos commented on frontends/web/src/routes/account/summary/summarybalance.tsx:
```diff
@@ -48,34 +44,25 @@ type TProps = {
   totalBalance?: accountApi.TAccountTotalBalance;
   balances?: Balances;
   keystoreDisambiguatorName?: string;
+  lightningBalance: accountApi.IBalance | undefined;
```
@thisconnect can this be simplified like this? same below in `TLightningProps`

```suggestion
  lightningBalance?: accountApi.IBalance;
```

thisconnect replied:
yes but with subtile difference. 

with this: `lightningBalance?: accountApi.IBalance;`  we tell TypeScript that the prop can be omitted.

with `lightningBalance: accountApi.IBalance | undefined;` the prop must be used but the value can be `undefined`

Beerosagos commented on frontends/web/src/routes/account/summary/accountssummary.tsx:
```diff
@@ -65,9 +68,73 @@ export const AccountsSummary = ({
+    keystores += 1;
+    keystoreDisambiguatorName = isAmbiguousName(
+      lightningKeystoreName,
+      lightningKeystoreName
+        ? [...accountsByKeystore, { keystore: { name: lightningKeystoreName } }]
+        : accountsByKeystore
```
You are computing this several times: here and for each `SummaryBalance`. You could simplify and make it more efficient by reusing it, right?

strmci replied:
I see, ok I will update it

strmci replied:
I created a new variable `allKeystores` which allowed me to get rid of `keystores` var and simplify the usage of `isAmbiguousName`

Beerosagos commented on frontends/web/src/routes/account/summary/filters.tsx:
```diff
@@ -23,12 +23,16 @@ export const Filters = ({
   onDisplayAll
 }: TChartFiltersProps) => {
   const { t } = useTranslation();
+  if (!showFiltersChartPercentageDiff) {
+    return null;
+  }
```
Since this property completely hides the whole component, is it really necessary? I think you could avoid creating it and just decide if including the filters at all in the parent component

Beerosagos reviewed (changes requested):
Left some new comments. @strmci, sorry for nitpicking a bit, but this part of the code is already quite messy and I think it is important to keep the changes as simple as possible

thisconnect commented on frontends/web/src/routes/account/summary/accountssummary.tsx:
```diff
@@ -65,9 +68,73 @@ export const AccountsSummary = ({
+    }
+  }, [fetchLightningBalance, getCoinsTotalBalance, lightningConfig]);
+
+  const lightningKeystoreName = useLoad(() => {
+    const account = lightningConfig.accounts[0];
+    return account
+      ? accountApi.keystoreName(account.rootFingerprint).then(response => response.keystoreName ?? '')
+      : Promise.resolve('');
+  }, [lightningConfig]) || '';
```
I agree it is a bit hard to read.

thisconnect commented on frontends/web/src/api/account.ts:
```diff
@@ -440,6 +440,16 @@ export const connectKeystore = (code: AccountCode): Promise<{ success: boolean;
+  keystoreName?: string;
+}
+
+export const keystoreName = (rootFingerprint: string): Promise<TResponseKeystoreName> => {
```
since you are here, ideally always return Fail | Success, that way keystoreName does not have to be optional since we know it is always a string in case of success. Also see other response types.

strmci commented:
@Beerosagos I made some updates:

- replaced `useLoad` for getting the lightning keystore name with `fetchLightningKeystoreName`
- removed unnecessary condition from  `getKeystoreName`
- added a check for `mounted.current` before calling `setLightningBalance`
- removed `showFiltersChartPercentageDiff` from `chartFiltersProps` and `filters.tsx` and updated conditions in `chart.tsx`
- reusing the `{ keystore: { name: lightningKeystoreName } } ` object
- comment and formatting updates

PTAL

Beerosagos commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -487,9 +488,10 @@ export const Chart = ({
 
   const hasDifference = difference && Number.isFinite(difference);
   const disableFilters = !hasData || chartDataMissing;
+  const showFiltersChartPercentageDiff: boolean = chartDataMissing || hasData || !!noDataPlaceholder;
```
I'm not sure if it wouldn't be better to add a new boolean prop to the Chart object, smth like "hideChartDetails" that we can set if the lightning account is the only one available, and that can be used instead of `showFiltersChartPercentageDiffs`. The reason is that it looks more maintainable to me, as it is clear what it does and why and it's easier to realize when it's not useful anymore and can be removed. But not a strong opinion, we can also leave it as is. cc @thisconnect wdyt?

strmci replied:
@Beerosagos I updated to use `hideChartDetails` as a `Chart` boolean prop.

Beerosagos commented on frontends/web/src/routes/account/summary/chart.tsx:
```diff
@@ -516,50 +519,56 @@ export const Chart = ({
-              </h2>
-              <span className={styles.toolTipTime}>
-                {renderDate(toolTipTime * 1000, i18n.language, source)}
+      {(!hideChartDetails) ? (
```
Nit: You could also do `{hideChartDetails || (...)}`

strmci replied:
Updated

Beerosagos reviewed (approved):
tACK! :partying_face: 
(with a last small nit)
